### PR TITLE
Fix broken CMake-generated pkg-config files

### PR DESCRIPTION
**Pull request by Artturin**
###### Description of changes
Cleaned up https://github.com/NixOS/nixpkgs/pull/172150
See https://github.com/NixOS/nixpkgs/pull/172150

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/radio/soapysdr/default.nix [hunk 1]
```diff
@@ -35,6 +35,13 @@ in stdenv.mkDerivation {
     "-DCMAKE_BUILD_TYPE=Release"
   ] ++ lib.optional usePython "-DUSE_PYTHON_CONFIG=ON";
 
+  # https://github.com/pothosware/SoapySDR/issues/352
+  postPatch = ''
+    substituteInPlace lib/SoapySDR.in.pc \
+      --replace '$'{exec_prefix}/@CMAKE_INSTALL_LIBDIR@ @CMAKE_INSTALL_FULL_LIBDIR@ \
```
Comment (Artturin) [inline pkgs/applications/radio/soapysdr/default.nix:41]: @alexshpilkin

`''${something}` is the proper way to escape `${`

If that is what you were trying to do then I'll change them all to that

Comment (alexshpilkin) [reply pkgs/applications/radio/soapysdr/default.nix:41]: No, that wasn’t it. This is more clever than I would like, but it has a purpose: there are two levels of escaping here, from Nix when it’s generating the script and from Bash when it’s executing it. That is to say, the file we’re patching contains the literal string `${exec_prefix}/@CMAKE_INSTALL_LIBDIR@` in it—the `${...}` is not shell syntax, it is pkg-config syntax. If you wanted to do this with standard Nix escaping as you suggest, you would end with `\''${exec_prefix}` or `'''${exec_prefix}'`, which is less subtle but even harder to parse I think.

(This is when .in files are being patched; take care that in some cases, CMake sources are being patched instead—so the interpolation is passing through Nix, Bash, and CMake on the way to the pkg-config file—and the CMake file being patched is already escaping the variable reference in one of a couple ways I’ve seen.)

If you see a clearer way, go for it, but the change you’re suggesting here wouldn’t work.

Comment (Artturin) [reply pkgs/applications/radio/soapysdr/default.nix:41]: Alright, do you think this is ready to merge?

**Review by fufexan — APPROVED** [review]
openal-soft LGTM

File: pkgs/development/libraries/drumstick/default.nix [hunk 2]
```diff
@@ -18,6 +18,20 @@ stdenv.mkDerivation rec {
...
+      --replace '$'{prefix}/@CMAKE_INSTALL_INCLUDEDIR@ @CMAKE_INSTALL_FULL_INCLUDEDIR@
+    substituteInPlace drumstick-file.pc.in \
+      --replace '$'{prefix}/@CMAKE_INSTALL_LIBDIR@ @CMAKE_INSTALL_FULL_LIBDIR@ \
+      --replace '$'{prefix}/@CMAKE_INSTALL_INCLUDEDIR@ @CMAKE_INSTALL_FULL_INCLUDEDIR@
+    substituteInPlace drumstick-rt.pc.in \
+      --replace '$'{prefix}/@CMAKE_INSTALL_LIBDIR@ @CMAKE_INSTALL_FULL_LIBDIR@ \
+      --replace '$'{prefix}/@CMAKE_INSTALL_INCLUDEDIR@ @CMAKE_INSTALL_FULL_INCLUDEDIR@
+    substituteInPlace drumstick-widgets.pc.in \
+      --replace '$'{prefix}/@CMAKE_INSTALL_LIBDIR@ @CMAKE_INSTALL_FULL_LIBDIR@ \
+      --replace '$'{prefix}/@CMAKE_INSTALL_INCLUDEDIR@ @CMAKE_INSTALL_FULL_INCLUDEDIR@
```
Comment (bjornfor) [inline pkgs/development/libraries/drumstick/default.nix:34]: I'd would have used a `for` here.

Comment (Artturin) [reply pkgs/development/libraries/drumstick/default.nix:34]: changed to passing multiple filenames https://github.com/NixOS/nixpkgs/pull/179603

**Comment by Artturin** [comment]
all of these were built successfully 

list generated with `git log upstream/staging..HEAD --oneline | cut -d' ' -f2 | sd ':' '' | uniq`

```
libiio
libzra
xtensor
xsimd
libcotp
libbaseencode
tinyobjloader
tdlib
spirv-headers
spdlog
soapysdr
shadowsocks-libev
sentencepiece
seexpr
rocm-thunk
rnp
rinutils
reproc
recastnavigation
rabbitmq-c
proj
powercap
orcania
openxr-loader
opendht
opencolorio
olm
obexftp
notcurses
nanomsg
mysocketw
maliit-framework
libunarr
libtsm
libtorrent-rasterbar
libsurvive
libquotient
libnats-c
libmodule
lxqt.lxqt-build-tools
libkeyfinder
libdnf
libcotp
libcork
libbtbb
libbaseencode
google-cloud-cpp
getdns
gbenchmark
entt
eccodes
plasma5Packages.drumstick
cxxopts
cog
cm256cc
cglm
bcg729
bcc
arrow-cpp
libargs
libebml
extra-cmake-modules
libmatroska
ffmpegthumbnailer
openobex
libebur128
openalSoft
usrsctp
wildmidi
libjxl
spirv-tools
gst_all_1.gst-plugins-bad
zxing
```

**Comment by Artturin** [comment]
So i am going to merge this

File: pkgs/tools/networking/shadowsocks-libev/default.nix [hunk 3]
```diff
@@ -25,6 +25,26 @@ stdenv.mkDerivation rec {
...
+    substituteInPlace cmake/shadowsocks-libev.pc.cmake \
+      --replace @prefix@ @CMAKE_INSTALL_PREFIX@ \
+      --replace '$'{prefix}/@CMAKE_INSTALL_BINDIR@ @CMAKE_INSTALL_FULL_BINDIR@ \
+      --replace '$'{exec_prefix}/@CMAKE_INSTALL_FULL_LIBDIR@ @CMAKE_INSTALL_FULL_LIBDIR@ \
+      --replace '$'{prefix}/@CMAKE_INSTALL_INCLUDEDIR@ @CMAKE_INSTALL_FULL_INCLUDEDIR@ \
+      --replace '$'{prefix}/@CMAKE_INSTALL_DATAROOTDIR@ @CMAKE_INSTALL_FULL_DATAROOTDIR@ \
+      --replace '$'{prefix}/@CMAKE_INSTALL_MANDIR@ @CMAKE_INSTALL_FULL_MANDIR@
+
+    # https://github.com/dcreager/libcork/issues/173 but needs a different patch (yay vendoring)
+    substituteInPlace libcork/src/libcork.pc.in \
```
Comment (jtojnar) [inline pkgs/tools/networking/shadowsocks-libev/default.nix:43]: Why would pc file of a vendored library be relevant?

Comment (Artturin) [reply pkgs/tools/networking/shadowsocks-libev/default.nix:43]: idk @alexshpilkin

**Comment by trofi** [comment]
`openalSoft` claims to fail patch hash after patch added in this PR. Does it work for you on `staging`?

```
$ nix build --no-link -f. openalSoft
...
error: hash mismatch in fixed-output derivation '/nix/store/2pmlpjzkd3yyayxkfnqhpl84mhbq9i7x-0bf2abae9b2121c3bc5a56dab30eca308136bc29.patch.drv':
         specified: sha256-QFio2LIEbF8VzLL8fRDRRusBkbtYV8qBGjLDbI+XEro=
            got:    sha256-IUm/5Aj8e1rTgvOL9wB08ilLWQXGPptfIRayMjkt2+A=
error (ignored): error: cannot unlink '/tmp/nix-build-util-linux-minimal-2.3
```

**Comment by Artturin** [comment]
> `openalSoft` claims to fail patch hash after patch added in this PR. Does it work for you on `staging`?
> 
> ```
> $ nix build --no-link -f. openalSoft
> ...
> error: hash mismatch in fixed-output derivation '/nix/store/2pmlpjzkd3yyayxkfnqhpl84mhbq9i7x-0bf2abae9b2121c3bc5a56dab30eca308136bc29.patch.drv':
>          specified: sha256-QFio2LIEbF8VzLL8fRDRRusBkbtYV8qBGjLDbI+XEro=
>             got:    sha256-IUm/5Aj8e1rTgvOL9wB08ilLWQXGPptfIRayMjkt2+A=
> error (ignored): error: cannot unlink '/tmp/nix-build-util-linux-minimal-2.3
> ```

Yeah but I removed the patch in https://github.com/NixOS/nixpkgs/pull/172347/commits/c9b34f760175584e0e77cd84fa2b97e0f060b57d

**Comment by trofi** [comment]
Oh, right. I completely missed. Will double-check locally if staging still fails.

**Pull request merged**